### PR TITLE
Overwrite error signals during backprop instead of adding to them

### LINKED ISSUE #347: Error signals overwrite matrices rather than update them
Note that this depends on moving to the layer-based objective functions

### PR DESCRIPTION
**Pull request by timmoon10**
I see ~8% speedup in AlexNet on Surface gpgpu, with no significant change to learning behavior. Gradient checks on model_zoo/tests/model_mnist_conv_graph.prototext come back clean. This closes #347 and is a step toward #193.
 
The kludgiest part of this PR involves the target layer. The objective function terms assume that the target layer's error signals are zeroed out when we compute their gradients, so the target layer's forward prop clears out the error signals. I've also added a default implementation of backprop which passes zero error signals. This makes it easier to naively implement layers (computed gradients will be inaccurate but not catastrophic).

Note that this breaks the VAE KL divergence objective function term, which is being removed in #431. It will probably also conflict with #405 and #417, so those should be merged first.

### CODE REVIEW

**Comment by bvanessen** [comment]
Tim can you please expand upon the statement: "This makes it easier to naively implement layers (computed gradients will be inaccurate but not catastrophic)."

**Comment by timmoon10** [comment]
If someone forgets to implement backprop, the computed gradient won't take the layer into account. If there were no default, the error signal would be filled with junk, maybe infs and nans.

File: include/lbann/layers/regularizers/selu_dropout.hpp [hunk 1]
```diff
@@ -136,7 +136,7 @@ class selu_dropout : public regularizer_layer {
   void bp_compute() override {
     if (this->m_model->get_execution_mode() != execution_mode::training
         || m_keep_prob < 0.0f) {
-      El::LockedView(get_error_signals(), get_prev_error_signals());
+      El::Copy(get_prev_error_signals(), get_error_signals());
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/regularizers/selu_dropout.hpp]: Good catch, I'll make that happen

**Comment by bvanessen** [comment]
Okay. By the way is it safe to assume that this optimization does not effect training.


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Jun 15, 2018, at 7:59 PM, Tim Moon <notifications@github.com<mailto:notifications@github.com>> wrote:


If someone forgets to implement backprop, the computed gradient won't take the layer into account. If there were no default, the error signal would be filled with junk, maybe infs and nans.

File: include/lbann/layers/regularizers/selu_dropout.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/regularizers/selu_dropout.hpp]: On second thought, it takes some slight convolutions to make the activations or error signals a view, even more when it changes between owning data during training and being a view during evaluation. The SeLU stuff is exotic and not yet performance critical (especially since this optimization would only speed up evaluation), so I think simplicity is more important.

**Comment by bvanessen** [comment]
Okay.


Brian C. Van Essen

vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>

(w) 925-422-9300

(c) 925-290-5470

---

Sent from my iPhone

On Jun 19, 2018, at 11:46 AM, Tim Moon <notifications@github.com<mailto:notifications@github.com>> wrote:


@timmoon10 commented on this pull request.

________________________________

In include/lbann/layers/regularizers/selu_dropout.hpp<https://github.com/LLNL/lbann/pull/435#discussion_r196537936>:

> @@ -136,7 +136,7 @@ class selu_dropout : public regularizer_layer {
   void bp_compute() override {
     if (this->m_model->get_execution_mode() != execution_mode::training
         || m_keep_prob < 0.0f) {
-      El::LockedView(get_error_signals(), get_prev_error_signals());
+      El::Copy(get_prev_error_signals(), get_error_signals());


On second thought, it takes some slight convolutions to make the activations or error signals a view, even more when it changes between owning data during training and being a view during evaluation. The SeLU stuff is exotic and not yet performance critical (especially since this optimization would only speed up evaluation), so I think simplicity is more important.
